### PR TITLE
feat: Istio multi primary single network blueprint

### LINKED ISSUE #1726: Create Istio multi-cluster/multi-primary setup blueprint
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

#### What is the outcome that you are trying to reach?

Create a complete blueprint for Istio multi-primary setup as described in [Istio Multi-Primary docs](https://istio.io/latest/docs/setup/install/multicluster/multi-primary/)
Searching the web came up with this [blog post](https://dev.to/redopsbay/setup-multi-cluster-servicemesh-with-istio-on-eks-5d5), but it automate part of the cluster creation.
A complete blueprint automating all the installation parts and trust set up can assist.

#### Describe the solution you would like

A fully automate blueprint (combined of several steps) to set up a multi-primary setup for Amazon EKS with Istio

#### Describe alternatives you have considered

This blog post https://dev.to/redopsbay/setup-multi-cluster-servicemesh-with-istio-on-eks-5d5

#### Additional context

I have a PR ready for that

### PR DESCRIPTION
**Pull request by tsahiduek**
# Description

Introducing a blueprint for multi-cluster Istio deployment in Multi-Primary setup

### Motivation and Context

- Resolves #1726 1726

### How was this change tested?

- [X] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [X] Yes, I have updated the [docs](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/docs) for this feature
- [ ] Yes, I ran `pre-commit run -a` with this PR

### Additional Notes

How should I run the `pre-commit run -a` command?

### CODE REVIEW

File: examples/istio-multi-cluster/0.certs-tool/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# Generating Certificates for Bootstrapping Multicluster / Mesh Expansion Chain of Trust
+
+The directory contains two Makefiles for generating new root, intermediate certificates and workload certificates:
+- `Makefile.k8s.mk`: Creates certificates based on a root-ca from a k8s cluster. The current context in the default
+`kubeconfig` is used for accessing the cluster.
+- `Makefile.selfsigned.mk`: Creates certificates based on a generated self-signed root.
```
Comment (rodrigobersa) [inline examples/istio-multi-cluster/0.certs-tool/README.md:6]: The directory contains two Makefiles for generating new root, intermediate and workload certificates:
- `Makefile.k8s.mk`: Creates certificates based on a root-ca from a Kubernetes cluster. The current context in the default `kubeconfig` is used to access the cluster.
- `Makefile.selfsigned.mk`: Creates certificates based on a generated self-signed root.

File: examples/istio-multi-cluster/0.vpc/main.tf [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+locals {
+  cluster_name = format("%s-%s", basename(path.cwd), "shared")
+  region       = "eu-west-1"
+
+  vpc_cidr = "10.0.0.0/16"
+  azs      = slice(data.aws_availability_zones.available.names, 0, 3)
+
+  # istio_chart_url     = "https://istio-release.storage.googleapis.com/charts"
+  # istio_chart_version = "1.18.1"
```
Comment (rodrigobersa) [inline examples/istio-multi-cluster/0.vpc/main.tf:15]: Remove commented lines.

File: examples/istio-multi-cluster/0.vpc/main.tf [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  }
+
+  private_subnet_tags = {
+    "kubernetes.io/role/internal-elb" = 1
+  }
+
+  tags = local.tags
+}
+
+
```
Comment (rodrigobersa) [inline examples/istio-multi-cluster/0.vpc/main.tf:52]: Remove one of the blank lines.

File: examples/istio-multi-cluster/0.vpc/outputs.tf [hunk 4]
```diff
@@ -0,0 +1,22 @@
+output "vpc_id" {
+  description = "Amazon EKS VPC ID"
+  value       = module.vpc.vpc_id
+}
```
Comment (rodrigobersa) [inline examples/istio-multi-cluster/0.vpc/outputs.tf:4]: Add a line after closing the output.

File: examples/istio-multi-cluster/0.vpc/outputs.tf [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+output "subnet_ids" {
+  description = "Amazon EKS Subnet IDs"
+  value       = module.vpc.private_subnets
+}
```
Comment (rodrigobersa) [inline examples/istio-multi-cluster/0.vpc/outputs.tf:8]: Same here.

File: examples/istio-multi-cluster/0.vpc/versions.tf [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    }
+  }
+
+
+  # ##  Used for end-to-end testing on project; update to suit your needs
+  # backend "s3" {
+  #   bucket = "terraform-ssp-github-actions-state"
+  #   region = "us-west-2"
+  #   key    = "e2e/istio/terraform.tfstate"
+  # }
```
Comment (rodrigobersa) [inline examples/istio-multi-cluster/0.vpc/versions.tf:17]: Remove commented lines.

File: examples/istio-multi-cluster/1.cluster1/README.md
Comment (rodrigobersa) [inline examples/istio-multi-cluster/1.cluster1/README.md:1]: Aren't there any steps that we can automate other than add instructions for customers to run.

File: examples/istio-multi-cluster/1.cluster1/outputs.tf [hunk 7]
```diff
@@ -0,0 +1,36 @@
+output "configure_kubectl" {
+  description = "Configure kubectl: make sure you're logged in with the correct AWS profile and run the following command to update your kubeconfig"
+  value       = "aws eks --region ${local.region} update-kubeconfig --name ${module.eks.cluster_name}"
+}
+
+output "istio_reader_token" {
+  description = "Istio token for cross cluster auth"
+  value       = data.kubernetes_secret.istio_reader.data["token"]
+  sensitive   = true
+}
```
Comment (rodrigobersa) [inline examples/istio-multi-cluster/1.cluster1/outputs.tf:10]: Add a blank line after closing the output.

File: examples/istio-multi-cluster/1.cluster1/outputs.tf [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  description = "Cluster ca certificate"
+  value       = module.eks.cluster_certificate_authority_data
+}
+
+output "cluster_name" {
+  description = "Cluster name"
+  value       = module.eks.cluster_name
+}
+
+
```
Comment (rodrigobersa) [inline examples/istio-multi-cluster/1.cluster1/outputs.tf:32]: Remove one of the blank lines.

File: examples/istio-multi-cluster/1.cluster1/main.tf [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+      }
+    )
+  ]
+}
+
+################################################################################
+# Isito certs for cross-cluster traffice
+# https://istio.io/latest/docs/ops/deployment/deployment-models/#trust-within-a-mesh
+# https://istio.io/latest/docs/ops/diagnostic-tools/multicluster/#trust-configuration
+################################################################################
```
Comment (rodrigobersa) [inline examples/istio-multi-cluster/1.cluster1/main.tf:232]: Add a line here, to keep format consistent.

File: examples/istio-multi-cluster/1.cluster1/versions.tf [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+    helm = {
+      source  = "hashicorp/helm"
+      version = ">= 2.9"
+    }
+    kubernetes = {
+      source  = "hashicorp/kubernetes"
+      version = ">= 2.20"
+    }
+  }
+
```
Comment (rodrigobersa) [inline examples/istio-multi-cluster/1.cluster1/versions.tf:18]: Remove this blank line.

File: examples/istio-multi-cluster/2.cluster2/README.md
Comment (rodrigobersa) [inline examples/istio-multi-cluster/2.cluster2/README.md:1]: Same here about automate steps or resources creation.

File: examples/istio-multi-cluster/2.cluster2/main.tf [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+      }
+    )
+  ]
+}
+
+################################################################################
+# Isito certs for cross-cluster traffice
+# https://istio.io/latest/docs/ops/deployment/deployment-models/#trust-within-a-mesh
+# https://istio.io/latest/docs/ops/diagnostic-tools/multicluster/#trust-configuration
+################################################################################
```
Comment (rodrigobersa) [inline examples/istio-multi-cluster/2.cluster2/main.tf:232]: New line here.

File: examples/istio-multi-cluster/2.cluster2/outputs.tf [hunk 12]
```diff
@@ -0,0 +1,35 @@
+output "configure_kubectl" {
+  description = "Configure kubectl: make sure you're logged in with the correct AWS profile and run the following command to update your kubeconfig"
+  value       = "aws eks --region ${local.region} update-kubeconfig --name ${module.eks.cluster_name}"
+}
+
+output "istio_reader_token" {
+  description = "Istio token for cross cluster auth"
+  value       = data.kubernetes_secret.istio_reader_data.data["token"]
+  sensitive   = true
+}
```
Comment (rodrigobersa) [inline examples/istio-multi-cluster/2.cluster2/outputs.tf:10]: Add a blank line after closing the output.

File: examples/istio-multi-cluster/2.cluster2/versions.tf [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+    helm = {
+      source  = "hashicorp/helm"
+      version = ">= 2.9"
+    }
+    kubernetes = {
+      source  = "hashicorp/kubernetes"
+      version = ">= 2.20"
+    }
+  }
+
```
Comment (rodrigobersa) [inline examples/istio-multi-cluster/2.cluster2/versions.tf:18]: Remove this line

File: examples/istio-multi-cluster/3.istio-multi-primary/main.tf [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+################################################################################
+
+# data "terraform_remote_state" "vpc" {
+#   backend = "local"
+
+#   config = {
+#     path = "${path.module}/../0.vpc/terraform.tfstate"
+#   }
+# }
+
```
Comment (rodrigobersa) [inline examples/istio-multi-cluster/3.istio-multi-primary/main.tf:17]: Remove commented lines.

File: examples/istio-multi-cluster/3.istio-multi-primary/versions.tf [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+    helm = {
+      source  = "hashicorp/helm"
+      version = ">= 2.9"
+    }
+    kubernetes = {
+      source  = "hashicorp/kubernetes"
+      version = ">= 2.20"
+    }
+  }
+
```
Comment (rodrigobersa) [inline examples/istio-multi-cluster/3.istio-multi-primary/versions.tf:18]: Remove this line

**Review by rodrigobersa — COMMENTED** [review]
Hey @tsahiduek 

Awesome job!! 

Just a a few formatting fixes. And a question about automate the validation steps.
Remember to run `pre-commit run --files` against each directory under your example.

File: examples/istio-multi-cluster/0.vpc/main.tf (hunk 2 above)
Comment (tsahiduek) [reply examples/istio-multi-cluster/0.vpc/main.tf:15]: done

File: examples/istio-multi-cluster/0.vpc/versions.tf (hunk 6 above)
Comment (tsahiduek) [reply examples/istio-multi-cluster/0.vpc/versions.tf:17]: this is for the E2E testing. It's the GH action that remove them when E2E is running. 
I fixed the path (key)

File: examples/istio-multi-cluster/1.cluster1/versions.tf (hunk 10 above)
Comment (tsahiduek) [reply examples/istio-multi-cluster/1.cluster1/versions.tf:18]: added remote-state for e2e testing

File: examples/istio-multi-cluster/2.cluster2/versions.tf (hunk 13 above)
Comment (tsahiduek) [reply examples/istio-multi-cluster/2.cluster2/versions.tf:18]: added remote state for e2e testing

**Review by rodrigobersa — APPROVED** [review]

**Pull request merged**